server/src/main/java/org/elasticsearch/cluster/coordination/NodeJoinExecutor.java

```diff
@@ -158,6 +159,7 @@ public ClusterState execute(BatchExecutionContext<JoinTask> batchExecutionContex
                 } else {
                     try {
                         CompatibilityVersions compatibilityVersions = nodeJoinTask.compatibilityVersions();
+                        assert systemIndexVersionConsistent(compatibilityVersions.systemIndexMappingsVersion(), compatibilityVersionsMap);
```

Comment:
Using assert means this won't run in production. Should we also ensure these match regardless of whether asserts are turned on?

Reply:
Definitely -- I put up this PR as a draft so I could run CI, work out a good test case, and figure out what we need to communicate to plugin authors. Once all that's done, I'll put this in an `ensureSystemIndexVersionConsistent` along the lines of the other checks in this block of code.